pkg/kubelet/server.go

```diff
@@ -620,81 +633,216 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+
+	glog.V(5).Infof("Upgrading port forward response")
 	upgrader := spdy.NewResponseUpgrader()
-	conn := upgrader.UpgradeResponse(response.ResponseWriter, request.Request, func(stream httpstream.Stream) error {
+	conn := upgrader.UpgradeResponse(w, req, portForwardStreamReceived(streamChan))
+	if conn == nil {
+		return
+	}
+	defer conn.Close()
+
+	glog.V(5).Infof("(conn=%p) setting port forwarding streaming connection idle timeout to %v", conn, idleTimeout)
+	conn.SetIdleTimeout(idleTimeout)
+
+	h := &portForwardStreamHandler{
+		conn:       conn,
+		streamChan: streamChan,
+		pairs:      make(map[string]*portForwardPair),
+		pod:        podName,
+		uid:        uid,
+		forwarder:  portForwarder,
+	}
+	h.run()
+}
+
+func portForwardStreamReceived(streams chan httpstream.Stream) func(httpstream.Stream) error {
+	return func(stream httpstream.Stream) error {
+		// make sure it has a valid port header
 		portString := stream.Headers().Get(api.PortHeader)
 		port, err := strconv.ParseUint(portString, 10, 16)
 		if err != nil {
-			return fmt.Errorf("Unable to parse '%s' as a port: %v", portString, err)
+			return fmt.Errorf("unable to parse %q as a port: %v", portString, err)
 		}
 		if port < 1 {
-			return fmt.Errorf("Port '%d' must be greater than 0", port)
+			return fmt.Errorf("port %d must be greater than 0", port)
+		}
+
+		// make sure it has a valid stream type header
+		streamType := stream.Headers().Get(api.StreamType)
+		if streamType != api.StreamTypeError && streamType != api.StreamTypeData {
+			return fmt.Errorf("invalid stream type %q", streamType)
 		}
-		streamChan <- stream
+
+		streams <- stream
 		return nil
-	})
-	if conn == nil {
-		return
 	}
-	defer conn.Close()
-	conn.SetIdleTimeout(s.host.StreamingConnectionIdleTimeout())
+}
+
+type portForwardStreamHandler struct {
+	conn       httpstream.Connection
+	streamChan chan httpstream.Stream
+	pairsLock  sync.Mutex
+	pairs      map[string]*portForwardPair
+	pod        string
+	uid        types.UID
+	forwarder  PortForwarder
+}
+
+func (h *portForwardStreamHandler) getPair(id string) *portForwardPair {
+	h.pairsLock.Lock()
+	defer h.pairsLock.Unlock()
+
+	if p, ok := h.pairs[id]; ok {
+		glog.V(5).Infof("(conn=%p) found existing session %s", h.conn, id)
+		return p
+	}
+
+	glog.V(5).Infof("(conn=%p) session %s is new", h.conn, id)
 
-	var dataStreamLock sync.Mutex
-	dataStreamChans := make(map[string]chan httpstream.Stream)
+	p := newPortForwardPair()
+	h.pairs[id] = p
 
+	go func() {
+		select {
+		case <-time.After(streamCreationTimeout):
+			util.HandleError(fmt.Errorf("(conn=%p, session=%s) timed out waiting for streams", h.conn, id))
+		case <-p.complete:
+			glog.V(5).Infof("(conn=%p, session=%s) successfully received error and data streams", h.conn, id)
+		}
+		h.removePair(id)
+	}()
+
+	return p
+}
+
+func (h *portForwardStreamHandler) removePair(id string) {
+	h.pairsLock.Lock()
+	defer h.pairsLock.Unlock()
+
+	delete(h.pairs, id)
```

Comment:
do we care about checking if the pair in the map for that id matches the one we're expecting to remove? not checking means recycled connection ids (which aren't a valid use) could get stomped